libnopegl/src/api.c

```diff
@@ -265,12 +265,13 @@ int ngli_ctx_set_scene(struct ngl_ctx *s, struct ngl_scene *scene)
             goto fail;
         }
 
+        s->scene = ngl_scene_ref(scene);
+
         ret = ngli_node_attach_ctx(scene->params.root, s);
         if (ret < 0) {
-            ngli_node_detach_ctx(scene->params.root, s);
-            return ret;
+            LOG(ERROR, "failed to attach scene");
+            goto fail;
         }
-        s->scene = ngl_scene_ref(scene);
     }
```

Comment:
Are you sure about this one? The attach can be partial and fail mid way after altering a bunch of nodes, so the detach seems useful. Did I miss something?

Reply:
The scene is detached after the fail label (in `reset_scene`). See:
``` c

fail:
    reset_scene(s, NGLI_ACTION_UNREF_SCENE);
    return ret;
```

``` c

static void reset_scene(struct ngl_ctx *s, int action)
{
    ngli_hud_freep(&s->hud);
    if (s->scene) {
        ngli_node_detach_ctx(s->scene->params.root, s);
        if (action == NGLI_ACTION_UNREF_SCENE)
            ngl_scene_unrefp(&s->scene);
    }
    ngli_rnode_reset(&s->rnode);
}
```